trace-utils/src/trace_utils.rs

```diff
@@ -0,0 +1,274 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present Datadog, Inc.
+
+use hyper::http::HeaderValue;
+use hyper::HeaderMap;
+use hyper::{body::Buf, Body, Client, Method, Request};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::{env, str};
+
+use prost::Message;
+
+use datadog_trace_protobuf::pb;
+
+const TRACE_INTAKE_URL: &str = "http://trace.agent.datadoghq.com/api/v0.2/traces";
+
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
+pub struct Span {
+    service: Option<String>,
+    name: String,
+    resource: String,
+    trace_id: u64,
+    span_id: u64,
+    parent_id: Option<u64>,
+    start: i64,
+    duration: i64,
+    error: Option<i32>,
+    meta: HashMap<String, String>,
+    metrics: Option<HashMap<String, f64>>,
+}
+
+pub async fn get_traces_from_request_body(body: Body) -> anyhow::Result<Vec<Vec<pb::Span>>> {
+    let buffer = hyper::body::aggregate(body).await.unwrap();
+
+    let traces: Vec<Vec<Span>> = match rmp_serde::from_read(buffer.reader()) {
+        Ok(res) => res,
+        Err(err) => {
+            anyhow::bail!("Error deserializing trace from request body: {}", err)
+        }
+    };
+
+    let mut pb_traces = Vec::<Vec<pb::Span>>::new();
+    for trace in traces {
+        let mut pb_spans = Vec::<pb::Span>::new();
+        for span in trace.iter() {
+            let span = pb::Span {
+                service: span.service.clone().unwrap_or_default(),
+                name: span.name.clone(),
+                resource: span.resource.clone(),
+                trace_id: span.trace_id,
+                span_id: span.span_id,
+                parent_id: span.parent_id.unwrap_or_default(),
+                start: span.start,
+                duration: span.duration,
+                error: span.error.unwrap_or(0),
+                meta: span.meta.clone(),
+                meta_struct: HashMap::new(),
+                metrics: span.metrics.clone().unwrap_or_default(),
+                r#type: "custom".to_string(),
+            };
+
+            pb_spans.push(span);
+        }
+        if !pb_spans.is_empty() {
+            pb_traces.push(pb_spans);
+        }
+    }
+
+    if pb_traces.is_empty() {
+        anyhow::bail!("No traces deserialized from the request body.")
+    }
+
+    Ok(pb_traces)
+}
+
+pub struct TracerTags<'a> {
+    lang: &'a str,
+    lang_version: &'a str,
+    lang_interpreter: &'a str,
+    lang_vendor: &'a str,
+    tracer_version: &'a str,
+}
+
+pub fn get_tracer_tags_from_request_header(headers: &HeaderMap<HeaderValue>) -> TracerTags {
+    let mut ts = TracerTags {
```

Comment:
maybe we could implement Default for TracerTags?

Reply:
done